### PR TITLE
8329661: Refactor ScavengableNMethods::verify_unlisted_nmethods

### PR DESCRIPTION
**Pull request by albertnetymk**
Simple removing redundant verification.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8329661](https://bugs.openjdk.org/browse/JDK-8329661): Refactor ScavengableNMethods::verify_unlisted_nmethods (**Enhancement** - P4)

### Reviewers
 * [Stefan Karlsson](https://openjdk.org/census#stefank) (@stefank - **Reviewer**)
 * [Ivan Walulya](https://openjdk.org/census#iwalulya) (@walulyai - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/18621/head:pull/18621` \
`$ git checkout pull/18621`

Update a local copy of the PR: \
`$ git checkout pull/18621` \
`$ git pull https://git.openjdk.org/jdk.git pull/18621/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 18621`

View PR using the GUI difftool: \
`$ git pr show -t 18621`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/18621.diff">https://git.openjdk.org/jdk/pull/18621.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/18621#issuecomment-2036921852)

### CODE REVIEW

**Comment by stefank** [comment]
Could you give a more detailed description why this is redundant. With stack traces and all?

I'm a little worried that this relies on the verify_nmethod call here:
```
void CodeCache::blobs_do(CodeBlobClosure* f) {
  assert_locked_or_safepoint(CodeCache_lock);
  FOR_ALL_ALLOCABLE_HEAPS(heap) {
    FOR_ALL_BLOBS(cb, *heap) {
      f->do_code_blob(cb);
#ifdef ASSERT
      if (cb->is_nmethod()) {
        Universe::heap()->verify_nmethod((nmethod*)cb);
      }
#endif //ASSERT
    }
  }
}
```

I'm not convinced that we should have that verification in the verification code and has been considering removing it.

**Comment by albertnetymk** [comment]
The real diff is in `ScavengableNMethods::verify_unlisted_nmethods`. The removed closure checks nmethods that are "not on-list" don't have old-to-young pointers. The same checking is already covered in `ScavengableNMethods::verify_nmethod`, which is invoked on all "live" (i.e. not unlinked) nmethods.

> I'm a little worried that this relies on the verify_nmethod call here:

In the Serial case, `ScavengableNMethods::prune_unlinked_nmethods` is used to cover the assert.

File: src/hotspot/share/gc/serial/serialHeap.cpp [hunk 1]
```diff
@@ -870,13 +855,6 @@ bool SerialHeap::is_in(const void* p) const {
 }
 
 #ifdef ASSERT
-// Don't implement this by using is_in_young().  This method is used
-// in some cases to check that is_in_young() is correct.
-bool SerialHeap::is_in_partial_collection(const void* p) {
-  assert(is_in_reserved(p) || p == nullptr,
-    "Does not work if address is non-null and outside of the heap");
-  return p < _young_gen->reserved().end() && p != nullptr;
-}
```
Comment (stefank) [inline src/hotspot/share/gc/serial/serialHeap.cpp:879]: I think the ifdefs should be removed as well.

**Review by stefank — COMMENTED** [review]
OK. I understand the change now. I wonder if it would make sense to rename `verify_unlisted_nmethods` now that you removed the section that visited "unlisted" nmethods. The function now only verifies "unlisted" nmethods inside `verify_nmethod` (which doesn't have the "_unlisted_" infix). Could this be named "verify_nmethods" instead?

**Review by walulyai — APPROVED** [review]

**Comment by albertnetymk** [comment]
Thanks for review.

/integrate

**Pull request closed without merging**